### ISSUE TITLE
Lazy expiration

### ISSUE THREAD

**Issue by y2k-shubham** [nektro/sdrie#5]
Problem
The current key-expiration implementation of `checkForDeadKeys()` has following limitations
 - Inconsistency: since it is based on **polling**, upon moving to *microsecond-precision expiration*, we will face issue where a key due to expire few microseconds earlier will still remain accessible until next delete cycle
 - Inefficiency: we have to keep running this background thread that does a lot of wasteful work: scanning over all keys in every cycle (irrespective of whether or not they were due to expire)
.
Proposed solution
**lazy expiration**
 - no need for background goroutine
 - at the time of access (`Has` or `Get`), check if key's expiration time has passed. And if it has, delete it (and accordingly change return value of `Has` / `Get`)

**Comment by nektro** [comment]
> Inconsistency

good point, and as I no longer actively use this in my own projects, this was not something I had considered to be used in such time critical scenarios.

> Inefficiency

per my comment in #2 only the first item in the list is checked, not the entire collection.

> lazy expiration

having this on top of the background solution would be ideal. using only this option would have expired keys that are never `Get`'d after they expire never having their memory freed.

**Comment by y2k-shubham** [comment]
@nektro , I've added lazy-cleanup of expired keys (and got rid of the cleanup goroutine). For cleaning up dead keys that aren't accessed in a while, I've set a threshold size for our map. Once the size of map (no of keys) grows beyond this threshold, a cleanup of all expired keys is triggered in the next operation (get, has, set or delete)
.
The obvious let-down here is that if all keys are meant to have big TTL, then all operations will slow down significantly once the size crosses the threshold
